src/DotNetProjectFile.Analyzers/NuGet/PackageCache.cs

```diff
@@ -0,0 +1,184 @@
+        var tildeExpanded = envExpanded.Replace("~", Environment.GetFolderPath(Environment.SpecialFolder.UserProfile).TrimEnd('/'));
+        var full = Path.GetFullPath(tildeExpanded);
+        return full;
+    }
+
+    private static string GetRawPath()
+    {
+        if (Environment.GetEnvironmentVariable("NUGET_PACKAGES") is { Length: > 0 } fromEnv)
+        {
+            return fromEnv;
+        }
+
+        // TODO: handle `globalPackagesFolder` given by `nuget.config`
+        // TODO: handle `repositoryPath` given by `packages.config`
+
+        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+        {
+            return @"%userprofile%\.nuget\packages";
+        }
+        else
+        {
+            return @"~/.nuget/packages";
+        }
+    }
+
+    public static CachedPackage? GetPackage(string? name, string? version)
+    {
+        if (name is not { Length: > 0 })
+        {
+            return null;
+        }
+
+        // Directories have lower case names.
+        name = name.ToLowerInvariant();
+        version = version?.ToLowerInvariant();
+
+        var key = (name, version);
+        if (cache.TryGetValue(key, out var result))
+        {
+            return result;
+        }
+
+        lock (cache)
+        {
+            if (!cache.TryGetValue(key, out result))
+            {
+                result = GetPackageInternal(name, version);
+
+                cache[key] = result;
+                
+                if (result is { } && version != result.Version)
+                {
+                    cache[(name, result.Version)] = result;
+                }
+            }
+        }
+
+        return result;
+    }
+
+    private static CachedPackage? GetPackageInternal(string name, string? version)
+    {
+        var cacheDir = GetPath();
+        var packageDir = Path.Combine(cacheDir, name);
+
+        if (!Directory.Exists(packageDir))
+        {
+            return null;
+        }
+
+        var (dir, foundVersion) = GetVersionDirectory(packageDir, version);
+
+        if (dir is null || foundVersion is null)
+        {
+            return null;
+        }
+
+        return new()
+        {
+            Name = name,
+            Version = foundVersion,
+            HasAnalyzerDll = HasDllFiles("analyzers"),
+            HasRuntimeDll = HasDllFiles("lib") || HasDllFiles("runtimes"),
+        };
+
+        bool HasDllFiles(string subDir)
+        {
+            var path = Path.Combine(dir, subDir);
+            if (!Directory.Exists(path))
+            {
+                return false;
+            }
+
+            return Directory.GetFiles(Path.Combine(dir, subDir), "*.dll", SearchOption.AllDirectories).Length > 0;
+        }
+    }
+
+    private static (string? Directory, string? Version) GetVersionDirectory(string packageDir, string? version)
+    {
+        // TODO: resolve version to a suitable fixed version when floating version is provided.
```

Comment:
Please add an issue, or resolve it in this PR.

Reply:
Created #320